framework/include/utils/Reconstructions.h

```diff
@@ -0,0 +1,103 @@
+ * between adjacent cell center values, and then creates an output functor whose cell-center
+ * evaluations will correspond to weighted averages of the input functor's surrounding face
+ * evaluations
+ * @param output_functor the output functor
+ * @param input_functor the input functor
+ * @param num_int_recs the total number of interpolation and reconstruction operations to perform.
+ * If this number is greater than 1, then this function will recurse
+ * @param two_term_expansion whether the cell center value reconstruction should consider the face
+ * gradients in addition to the face values
+ * @param weight_with_sf when reconstructing the cell center value, decides whether the face values
+ * (and maybe gradients) are weighted with the surface vector. If this is false, then the weights
+ * are simply unity
+ * @param faces the mesh faces we will be looping over for the interpolations and reconstructions
+ * @param consumer the object that needs the reconstructed field. This argument is useful for
+ * determining what are "external" faces and hence faces around which we should carefully choose the
+ * subdomains we want to evaluate our \p input_functor on
+ */
+template <typename T, typename Map, typename Consumer>
+void
+interpolateReconstruct(CellCenteredMapFunctor<T, Map> & output_functor,
+                       const Moose::FunctorBase<T> & input_functor,
+                       const unsigned int num_int_recs,
+                       const bool two_term_expansion,
+                       const bool weight_with_sf,
+                       const std::vector<const FaceInfo *> & faces,
+                       const Consumer & consumer)
+{
+  if (!num_int_recs)
+    return;
+
+  std::unordered_map<dof_id_type, std::pair<T, Real>> elem_to_num_denom;
+
+  for (const auto * const face : faces)
+  {
+    mooseAssert(face, "This must be non-null");
+    const Real weight = weight_with_sf ? face->faceArea() * face->faceCoord() : 1;
+    const auto sub_pair = faceArgSubdomains(consumer, *face);
+    const auto face_arg = makeCDFace(*face, sub_pair);
+    auto face_value = input_functor(face_arg);
+    std::pair<T, Real> * neighbor_pr = nullptr;
+    if (face->neighborPtr() && face->neighborPtr() != libMesh::remote_elem)
+    {
+      neighbor_pr = &elem_to_num_denom[face->neighbor().id()];
+      neighbor_pr->first += face_value * weight;
+      neighbor_pr->second += weight;
+    }
+    auto & elem_pr = elem_to_num_denom[face->elem().id()];
+    elem_pr.first += std::move(face_value) * weight;
+    elem_pr.second += weight;
+
+    if (two_term_expansion)
+    {
+      auto face_gradient = input_functor.gradient(face_arg);
+      if (face->neighborPtr() && face->neighborPtr() != libMesh::remote_elem)
+        neighbor_pr->first +=
+            face_gradient * (face->neighborCentroid() - face->faceCentroid()) * weight;
+      elem_pr.first +=
+          std::move(face_gradient) * (face->elemCentroid() - face->faceCentroid()) * weight;
+    }
```

Comment:
We are going the other way here. From the face to the cell.

Reply:
ooooh. Sorry about that. I need to review my reviews, this is embarrassing. 
Interesting alternative to average the "face" contributions. Seems expensive though.

Reply:
What do you mean by alternative? In one loop we are doing interpolation to faces and then reconstruction back to the cell center, so two overbar operations in one. I would say that's pretty good! If you're referring to something specific to use of the gradient ... I don't actually anticipate the `two_term_expansion` ever being used in practice. It actually perfectly reconstructs the original cell center value ... which is not what we want when we're trying to smooth a field.

Reply:
Which makes me think that I should probably remove the `two_term_expansion` option. It was nice for testing, but what's the point of just re-creating the field you already have? It would make sense in a context of *just* doing a reconstruction but not when doing successive interpolate and reconstruct.